conan.tools.microsoft.VCVars generator does not follow settings.compiler.update

[GenericSystemBlock.get_toolset](https://github.com/conan-io/conan/blob/8f86d38b92c22f82ad674ed69a1bf684fe056b32/conan/tools/cmake/toolchain/blocks.py#L696-L700) already has code to combine settings.compiler.version and settings.compiler.update such that compiler.version=193 and compiler.update=7 into a CMakePresets result like

```json
            "toolset": {
                "value": "version=14.37",
                "strategy": "external"
            },
```
(this doesn't quite work because VSCode also wants to see CMAKE_C_COMPILER=cl, but that's already reported in #13136).

However, [conan.tools.microsoft.visual._vcvars_vers](https://github.com/conan-io/conan/blob/8f86d38b92c22f82ad674ed69a1bf684fe056b32/conan/tools/microsoft/visual.py#L290) does not seem to have similar code to consider settings.compiler.update, so generators="VCVars" will only generate calls like `-vcvars_ver=14.3`, never `-vcvars_ver=14.37` vs `-vcvars_ver=14.38`. Which means conanbuild.bat won't necessarily get the same minor version that CMakePresets does... or even one that CMake will accept 

i.e.
```sh
conan new cmake_exe -d name=hello -d version=1.0
conan create . -s compiler.update=7
```

will fail with

> conanvcvars.bat: Activating environment Visual Studio 17 - amd64 - winsdk_version=None - vcvars_ver=14.3
> [vcvarsall.bat] Environment initialized for: 'x64'
> -- Using Conan toolchain: C:/Users/re41236/.conan2/p/b/hello23b22256ffb21/b/build/generators/conan_toolchain.cmake
> -- Conan toolchain: CMAKE_GENERATOR_TOOLSET=version=14.37
> -- Conan toolchain: C++ Standard 14 with extensions OFF
> -- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.19045.
> -- The CXX compiler identification is unknown
> CMake Error at CMakeLists.txt:2 (project):
>   No CMAKE_CXX_COMPILER could be found.

```sh
conan2 create . -s compiler.update=7 -c tools.cmake.cmaketoolchain:generator=Ninja
```

is even worse - it just uses compiler.update=8 despite the settings:
> The CXX compiler identification is MSVC 19.38.33134.0

In both cases this is because the conanvcvars.bat contains
> C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC/Auxiliary/Build/vcvarsall.bat"  amd64 -vcvars_ver=14.3
rather than
> C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC/Auxiliary/Build/vcvarsall.bat"  amd64 -vcvars_ver=14.37

So the environment finds the latest 14.3x (which is 14.38) and then CMake rejects it as not mathching CMAKE_GENERATOR_TOOLSET

Since conan_toolchain.cmake contains `set(CMAKE_GENERATOR_TOOLSET "version=14.37" CACHE STRING "" FORCE)`, the msbuild generator knows this is wrong; ninja (as usual) doesn't actually check, and just goes by PATH.

Similarly, a call like `conan2 create . -s compiler.version=192 -s compiler.update=9 -c tools.microsoft.msbuild:vs_version=17`, which should select C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.29.30133, failed with a CMake error:

_Originally posted by @puetzk in https://github.com/conan-io/conan/issues/13136#issuecomment-1909152616_

Hi @puetzk 

Thanks for reporting.
I am having a look, but I have some questions.

The above steps:
```
conan new cmake_exe -d name=hello -d version=1.0
conan create . -s compiler.update=8
```
Uses by default the ``Visual Studio`` generator, not the ``Ninja`` one, so the ``VCVars`` won't be used at all, but I guess that in your setup you have configured it.

However, I am trying to make it work with the ``Visual Studio`` CMake generator, as a first step to understand the issue, but:
- Work if ``compiler.update`` is the latest and active one installed, in my case ``8`` and it will use in the toolchain ``-- Conan toolchain: CMAKE_GENERATOR_TOOLSET=version=14.38``, and all works.
- I have installed in my VS 2022 a previous 14.36 update and now I am trying the same, but it fails with: 
```
-- Conan toolchain: CMAKE_GENERATOR_TOOLSET=version=14.36
-- Conan toolchain: C++ Standard 14 with extensions OFF
-- Selecting Windows SDK version 10.0.20348.0 to target Windows 10.0.19045.
-- The CXX compiler identification is unknown
CMake Error at CMakeLists.txt:2 (project):
  No CMAKE_CXX_COMPILER could be found.
```

Not sure what would be missing or necessary, maybe CMake cannot properly use previous updates? I have also tried hardcoding the full ``14.36.17.6`` toolset version, but same result

>     * I have installed in my VS 2022 a previous 14.36 update and now I am trying the same, but it fails with:
> 
> 
> ```
> -- Conan toolchain: CMAKE_GENERATOR_TOOLSET=version=14.36
> -- Conan toolchain: C++ Standard 14 with extensions OFF
> -- Selecting Windows SDK version 10.0.20348.0 to target Windows 10.0.19045.
> -- The CXX compiler identification is unknown
> CMake Error at CMakeLists.txt:2 (project):
>   No CMAKE_CXX_COMPILER could be found.
> ```
> 
> Not sure what would be missing or necessary, maybe CMake cannot properly use previous updates? I have also tried hardcoding the full `14.36.17.6` toolset version, but same result

This is probably caused by bug in Visual Studio that should be fixed in 17.10
https://developercommunity.visualstudio.com/t/MicrosoftVCToolsVersion1436176prop/10385615

--- a/conan/tools/microsoft/visual.py
+++ b/conan/tools/microsoft/visual.py
@@ -133,7 +133,13 @@ class VCVars:
                           "v143": "14.3"}.get(toolset_version)
         else:
             vs_version = vs_ide_version(conanfile)
-            vcvars_ver = _vcvars_vers(conanfile, compiler, vs_version)
+            if int(vs_version) <= 14:
+                vcvars_ver = None
+            else:
+                compiler_version = str(conanfile.settings.compiler.version)
+                compiler_update = conanfile.settings.get_safe("compiler.update", "")
+                # The equivalent of compiler 19.26 is toolset 14.26
+                vcvars_ver = "14.{}{}".format(compiler_version[-1], compiler_update)
         vcvarsarch = _vcvars_arch(conanfile)
 
         winsdk_version = conanfile.conf.get("tools.microsoft:winsdk_version", check_type=str)
@@ -177,7 +183,6 @@ class VCVars:
             _create_deactivate_vcvars_file(conanfile, conan_vcvars_ps1)
 
 
-
 def _create_deactivate_vcvars_file(conanfile, filename):
     deactivate_filename = f"deactivate_{filename}"
     message = f"[{deactivate_filename}]: vcvars env cannot be deactivated"
@@ -188,6 +193,7 @@ def _create_deactivate_vcvars_file(conanfile, filename):
         content = f"echo {message}"
     path = os.path.join(conanfile.generators_folder, deactivate_filename)
     save(path, content)
+
 
 def vs_ide_version(conanfile):
     """
@@ -298,7 +304,7 @@ def _vcvars_arch(conanfile):
                 'x86_64': 'amd64',
                 'armv7': 'amd64_arm',
                 'armv8': 'amd64_arm64',
-                'arm64ec':'amd64_arm64'}.get(arch_host)
+                'arm64ec': 'amd64_arm64'}.get(arch_host)
     elif arch_build == 'x86':
         arch = {'x86': 'x86',
                 'x86_64': 'x86_amd64',
@@ -314,18 +320,6 @@ def _vcvars_arch(conanfile):
         raise ConanException('vcvars unsupported architectures %s-%s' % (arch_build, arch_host))
 
     return arch
-
-
-def _vcvars_vers(conanfile, compiler, vs_version):
-    if int(vs_version) <= 14:
-        return None
-
-    assert compiler == "msvc"
-    # Code similar to CMakeToolchain toolset one
-    compiler_version = str(conanfile.settings.compiler.version)
-    # The equivalent of compiler 192 is toolset 14.2
-    vcvars_ver = "14.{}".format(compiler_version[-1])
-    return vcvars_ver
 
 
 def is_msvc(conanfile, build_context=False):

--- a/conans/test/integration/toolchains/microsoft/vcvars_test.py
+++ b/conans/test/integration/toolchains/microsoft/vcvars_test.py
@@ -51,6 +51,7 @@ def test_vcvars_generator_skip():
     client.run('install . -pr=profile')
     assert not os.path.exists(os.path.join(client.current_folder, "conanvcvars.bat"))
 
+
 @pytest.mark.skipif(platform.system() not in ["Linux"], reason="Requires Linux")
 def test_vcvars_generator_skip_on_linux():
     """
@@ -64,6 +65,7 @@ def test_vcvars_generator_skip_on_linux():
     client.run('install . -s os=Windows -s compiler=msvc -s compiler.version=193 '
                '-s compiler.runtime=dynamic')
     assert not os.path.exists(os.path.join(client.current_folder, "conanvcvars.bat"))
+
 
 @pytest.mark.skipif(platform.system() not in ["Windows"], reason="Requires Windows")
 def test_vcvars_generator_string():
@@ -140,6 +142,26 @@ def test_vcvars_winsdk_version():
     vcvars = client.load("conanvcvars.bat")
     assert 'vcvarsall.bat"  amd64 8.1 -vcvars_ver=14.3' in vcvars
 
+
+@pytest.mark.skipif(platform.system() != "Windows", reason="Requires Windows")
+def test_vcvars_compiler_update():
+    client = TestClient(path_with_spaces=False)
+
+    conanfile = textwrap.dedent("""
+        from conan import ConanFile
+        class TestConan(ConanFile):
+            generators = "VCVars"
+            settings = "os", "compiler", "arch", "build_type"
+    """)
+    client.save({"conanfile.py": conanfile})
+    client.run('install . -s os=Windows -s compiler=msvc -s compiler.version=193 '
+               '-s compiler.cppstd=14 -s compiler.runtime=static '
+               '-s compiler.update=3')
+
+    vcvars = client.load("conanvcvars.bat")
+    assert 'vcvarsall.bat"  amd64 -vcvars_ver=14.33' in vcvars
+
+
 @pytest.mark.skipif(platform.system() != "Windows", reason="Requires Windows")
 def test_deactivate_vcvars_message():
     client = TestClient()
@@ -155,6 +177,7 @@ def test_deactivate_vcvars_message():
     assert "[vcvarsall.bat] Environment initialized" in client.out
     client.run_command(r'deactivate_conanvcvars.bat')
     assert "vcvars env cannot be deactivated" in client.out
+
 
 @pytest.mark.skipif(platform.system() != "Windows", reason="Requires Windows Powershell")
 def test_deactivate_vcvars_with_powershell():

--- a/conans/test/unittests/tools/cmake/test_cmaketoolchain.py
+++ b/conans/test/unittests/tools/cmake/test_cmaketoolchain.py
@@ -195,7 +195,7 @@ def conanfile_msvc():
     c = ConanFile(None)
     c.settings = Settings({"os": ["Windows"],
                            "compiler": {"msvc": {"version": ["193"], "cppstd": ["20"],
-                                                 "update": [None]}},
+                                                 "update": [None, 8, 9]}},
                            "build_type": ["Release"],
                            "arch": ["x86"]})
     c.settings.build_type = "Release"
@@ -218,6 +218,12 @@ def test_toolset(conanfile_msvc):
     assert 'set(CMAKE_GENERATOR_TOOLSET "v143" CACHE STRING "" FORCE)' in toolchain.content
     assert 'Visual Studio 17 2022' in toolchain.generator
     assert 'CMAKE_CXX_STANDARD 20' in toolchain.content
+
+
+def test_toolset_update_version(conanfile_msvc):
+    conanfile_msvc.settings.compiler.update = "8"
+    toolchain = CMakeToolchain(conanfile_msvc)
+    assert 'set(CMAKE_GENERATOR_TOOLSET "v143,version=14.38" CACHE STRING "" FORCE)' in toolchain.content
 
 
 def test_toolset_x64(conanfile_msvc):

